File: local/app/game/tetris_cpp/src/scorer.cpp
#include "scorer.h"

Scorer::Scorer() { reset(); }

void Scorer::reset() {
    score = 0;
    level = 1;
    lines = 0;
}

void Scorer::add_clear(int n) {
    if (n <= 0 || n > 4) return;
    score += SCORE_TABLE[n] * level;
    lines += n;
    // 每 10 行升一级
    int new_level = lines / LINES_PER_LEVEL + 1;
    if (new_level > MAX_LEVEL) new_level = MAX_LEVEL;
    level = new_level;
}

void Scorer::add_soft_drop(int cells) { score += cells; }

void Scorer::add_hard_drop(int cells) { score += cells * 2; }

int Scorer::drop_speed_ms() const {
    return SPEED_TABLE[level];
}

File: local/app/game/tetris/scorer.py
"""计分 & 等级管理"""
from constants import SCORE_TABLE, LINES_PER_LEVEL, MAX_LEVEL, SPEED_TABLE


class Scorer:
    def __init__(self):
        self.score = 0
        self.level = 1
        self.lines = 0

    def add_clear(self, lines_cleared: int):
        if lines_cleared < 1:
            return
        self.score += SCORE_TABLE.get(lines_cleared, 800) * self.level
        self.lines += lines_cleared
        self.level = min(MAX_LEVEL, 1 + self.lines // LINES_PER_LEVEL)

    def add_soft_drop(self, cells: int = 1):
        self.score += cells

    def add_hard_drop(self, cells: int):
        self.score += cells * 2

    def drop_speed_ms(self) -> int:
        return SPEED_TABLE.get(self.level, 100)

    def reset(self):
        self.score = 0
        self.level = 1
        self.lines = 0
